### PR TITLE
PLT-2241 Refactored statuses into a more real-time system

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
* Removed LastAcitivityAt, LastPingAt from User model and database table
* Created Status model and database table
* Added in-memory status cache similar to sessions
* Created api/status.go for handling setting of statuses
* Modified websocket handlers to return response data instead of modifying a passed in parameter
* Added away timeout to config.json (not to System Console yet)

#### Ticket Link
* Completes https://mattermost.atlassian.net/browse/PLT-2241
* Fixes https://mattermost.atlassian.net/browse/PLT-3604
* Fixes https://mattermost.atlassian.net/browse/PLT-3545

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] Has driver changes that have been merged and package.json updated (mattermost/mattermost-driver-javascript#18)
- [x] Includes text changes and localization files updated
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: api/status.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	}
+}
+
+func SetStatusOnline(userId string, sessionId string) {
+	broadcast := false
+
+	var status *model.Status
+	var err *model.AppError
+	if status, err = GetStatus(userId); err != nil {
+		status = &model.Status{userId, model.STATUS_ONLINE, model.GetMillis()}
```
Comment (enahum) [inline api/status.go]: If you get an error here why the user is online? I don't get it

**Comment by enahum** [comment]
+1 looks good to me

**Comment by lfbrock** [comment]
Tested changing between online, offline, and away.

Only found one thing - if someone goes offline and you aren't in the Direct Message channel, the offline indicator disappears from the channel header. If they go offline when you are in the channel, the indicator in the header is there but it disappears when you refresh. 
screenshot 2016-07-13 13 14 41

Other than that +1, everything else worked fine

**Comment by lfbrock** [comment]
Looks good, +1

**Comment by jwilander** [comment]
Rebased and tested to make sure `away` status works correctly.

**Comment by enahum** [comment]
+1 seems ok to me, I can't tell the difference but it seems that now the `SetStatusAwayIfNeeded` is being called when the WebSocket pings, im i right?

**Comment by jwilander** [comment]
There is no difference except a minor client-side bug :) That is correct, the ping triggers a check to see if the user is away, it's actually very similar to how we were doing it before

File: api/status.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+			broadcast = true
+		}
+		status.Status = model.STATUS_ONLINE
+		status.LastActivityAt = model.GetMillis()
+	}
+
+	statusCache.Add(status.UserId, status)
+
+	achan := Srv.Store.Session().UpdateLastActivityAt(sessionId, model.GetMillis())
+	schan := Srv.Store.Status().SaveOrUpdate(status)
```
Comment (crspeller) [inline api/status.go]: If we discovered the status is already online above why do we bother saving anyways?

Comment (jwilander) [inline api/status.go]: To update the LastActivityAt field on the object. I could split it out to only use a database call to update the LastActivityAt but that would make the logic here a bit more complex and I figured it was better to keep it simpler

File: api/status.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	status, err := GetStatus(userId)
+	if err != nil {
+		status = &model.Status{userId, model.STATUS_OFFLINE, 0}
+	}
+
+	if status.Status == model.STATUS_AWAY {
+		return
+	}
+
+	if model.GetMillis()-status.LastActivityAt <= *utils.Cfg.TeamSettings.UserStatusAwayTimeout*1000 {
```
Comment (crspeller) [inline api/status.go]: Can we move this logic to a function. Something like `isUserConsiteredAway(lastActivity int)`

File: mattermost.go [hunk 4]
```diff
@@ -134,6 +134,7 @@ func main() {
 
 		setDiagnosticId()
 		go runSecurityAndDiagnosticsJob()
+		go resetStatuses()
```
Comment (crspeller) [inline mattermost.go]: Maybe this should run synchronous before the api.StartServer? What if a user connects before this is run?

File: store/sql_channel_store.go [hunk 5]
```diff
@@ -992,7 +992,7 @@ func (s SqlChannelStore) GetForExport(teamId string) StoreChannel {
 		_, err := s.GetReplica().Select(&data, "SELECT * FROM Channels WHERE TeamId = :TeamId AND DeleteAt = 0 AND Type = 'O'", map[string]interface{}{"TeamId": teamId})
 
 		if err != nil {
-			result.Err = model.NewLocAppError("SqlChannelStore.GetAllChannels", "store.sql_channel.get_for_export.app_error", nil, "teamId="+teamId+", err="+err.Error())
+			result.Err = model.NewLocAppError("SqlChannelStore.GetForExport", "store.sql_channel.get_for_export.app_error", nil, "teamId="+teamId+", err="+err.Error())
```
Comment (crspeller) [inline store/sql_channel_store.go]: This PR: https://github.com/mattermost/platform/pull/3601 deletes this function. Revert this change to avoid conflict.

File: api/status.go (hunk 2 above)
Comment (crspeller) [inline api/status.go]: I think it's a worthwhile optimization. It's not too complicated and it's a highly used piece of code. Also from working on LDAP sync I know that DB writes in our system are very expensive.

File: store/sql_channel_store.go (hunk 5 above)
Comment (jwilander) [inline store/sql_channel_store.go]: Cool, will do

File: mattermost.go (hunk 4 above)
Comment (jwilander) [inline mattermost.go]: It should be remedied pretty quick as soon as they click on something in the app, but to be more correct and accurate I can move it

File: api/status.go (hunk 3 above)
Comment (jwilander) [inline api/status.go]: Yeah makes sense, will do

File: api/status.go (hunk 2 above)
Comment (jwilander) [inline api/status.go]: Well we need to do a write to the Status table either way, is an update to only the LastActivityAt column much better than writing to all three columns?

**Comment by jwilander** [comment]
Ready for re-review @crspeller

**Comment by crspeller** [comment]
+1
